jxl/src/image.rs

```diff
@@ -0,0 +1,257 @@
+use crate::error::{Error, Result};
+
+mod private {
+    pub trait Sealed {}
+}
+
+pub trait ImageDataType: private::Sealed + Copy + Default {
+    fn data_type_id() -> usize;
+}
+
+macro_rules! impl_image_data_type {
+    ($ty: ty, $id: literal) => {
+        impl private::Sealed for $ty {}
+        impl ImageDataType for $ty {
+            fn data_type_id() -> usize {
+                $id
+            }
+        }
+    };
+}
+
+impl_image_data_type!(u8, 0);
+impl_image_data_type!(u16, 1);
+impl_image_data_type!(u32, 2);
+impl_image_data_type!(f32, 3);
+impl_image_data_type!(i8, 4);
+impl_image_data_type!(i16, 5);
+impl_image_data_type!(i32, 6);
+impl_image_data_type!(half::f16, 7);
+
+pub struct Image<T: ImageDataType> {
+    size: (usize, usize),
+    data: Vec<T>,
+}
+
+#[derive(Clone, Copy)]
+pub struct ImageRect<'a, T: ImageDataType> {
+    origin: (usize, usize),
+    size: (usize, usize),
+    image: &'a Image<T>,
+}
+
+pub struct ImageRectMut<'a, T: ImageDataType> {
+    origin: (usize, usize),
+    size: (usize, usize),
+    image: &'a mut Image<T>,
+}
+
+impl<T: ImageDataType> Image<T> {
+    pub fn new(xsize: usize, ysize: usize) -> Result<Image<T>> {
+        let total_size = xsize
+            .checked_mul(ysize)
+            .ok_or(Error::ImageSizeTooLarge(xsize, ysize))?;
+        let mut data = vec![];
+        data.try_reserve_exact(total_size)?;
+        data.resize(total_size, T::default());
+        Ok(Image {
+            size: (xsize, ysize),
+            data,
+        })
+    }
+
+    pub fn size(&self) -> (usize, usize) {
+        self.size
+    }
+
+    pub fn as_rect(&self) -> ImageRect<'_, T> {
+        ImageRect {
+            origin: (0, 0),
+            size: self.size,
+            image: self,
+        }
+    }
+
+    pub fn as_rect_mut(&mut self) -> ImageRectMut<'_, T> {
+        ImageRectMut {
+            origin: (0, 0),
+            size: self.size,
+            image: self,
+        }
+    }
+}
+
+fn rect_size_check(
+    origin: (usize, usize),
+    size: (usize, usize),
+    ssize: (usize, usize),
+) -> Result<()> {
+    if origin
+        .0
+        .checked_add(size.0)
+        .ok_or(Error::ArithmeticOverflow)?
+        > ssize.0
+        || origin
+            .1
+            .checked_add(size.1)
+            .ok_or(Error::ArithmeticOverflow)?
+            > ssize.1
+    {
+        Err(Error::RectOutOfBounds(
+            size.0, size.1, origin.0, origin.1, ssize.0, ssize.1,
+        ))
+    } else {
```

Comment:
no else needed here, just return `OK(())`?

Reply:
else is needed to avoid a return above... Doesn't make too much of a difference IMO